mllib/src/main/scala/org/apache/spark/ml/feature/TargetEncoder.scala

```
@@ -0,0 +1,442 @@
+ */
+@Since("4.0.0")
+class TargetEncoder @Since("4.0.0") (@Since("4.0.0") override val uid: String)
+  extends Estimator[TargetEncoderModel] with TargetEncoderBase with DefaultParamsWritable {
+
+  @Since("4.0.0")
+  def this() = this(Identifiable.randomUID("TargetEncoder"))
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setLabelCol(value: String): this.type = set(labelCol, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setInputCols(values: Array[String]): this.type = set(inputCols, values)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setOutputCols(values: Array[String]): this.type = set(outputCols, values)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setHandleInvalid(value: String): this.type = set(handleInvalid, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setTargetType(value: String): this.type = set(targetType, value)
+
+  /** @group setParam */
+  @Since("4.0.0")
+  def setSmoothing(value: Double): this.type = set(smoothing, value)
+
+  @Since("4.0.0")
+  override def transformSchema(schema: StructType): StructType = {
+    validateSchema(schema, fitting = true)
+  }
+
+  @Since("4.0.0")
+  override def fit(dataset: Dataset[_]): TargetEncoderModel = {
+    validateSchema(dataset.schema, fitting = true)
+
+    val feature_types = inputFeatures.map{
+      feature => dataset.schema(feature).dataType
+    }
+    val label_type = dataset.schema($(labelCol)).dataType
+
+    val stats = dataset
+      .select((inputFeatures :+ $(labelCol)).map(col).toIndexedSeq: _*)
+      .rdd.treeAggregate(
+        Array.fill(inputFeatures.length) {
+          Map.empty[Option[Double], (Double, Double)]
+        })(
+        (agg, row: Row) => if (!row.isNullAt(inputFeatures.length)) {
```

Ah, if the label is null, hm, that could be an error. I don't think it makes sense to have input to any supervised problem where the label is missing. Here you'd be welcome to copy whatever other similar classes do in this situation

---

i think we didn't understand each other
so, you want to raise an exception in case a label is null